### PR TITLE
Pytorch synthesizer

### PR DESCRIPTION
**Pull request by blue-fish**
I have taken the tacotron model from [fatchord/WaveRNN](https://github.com/fatchord/WaveRNN) and integrated it with this repo (#447). Aside from the new format of the synthesizer model (.pt) this change should be completely transparent to the end user.

#### Major Changes
* Toolbox no longer requires tensorflow 🎉  
* Synthesizer is tacotron1 instead of tacotron2

#### Pretrained Model
A download link and instructions are provided here: https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472#issuecomment-695206377

#### Task List
- [x] Inference
- [x] Training
- [x] Update preprocessing scripts to use synthesizer_pt
- [x] Cleanup files in synthesizer_pt
- [x] Match repo code style
- [x] Move synthesizer_pt to synthesizer (no more tensorflow)
- [x] Testing
- [x] Release pretrained model
- [x] Update documentation
- [x] Review
    - [x] Retest as needed if code changes made
- [ ] Merge into master branch of repo

### CODE REVIEW

**Comment by blue-fish** [comment]
@CorentinJ This is a huge PR and merging the hparams will make it even more difficult to review. Please do an initial review and we can either add it before the final review, or do a new PR once this has been merged.

We are still missing a pretrained model. The one I have works well for small utterances but does not align for longer ones, leading to large gaps in the output. I am training a new model to see if I can fix the issue.

**Comment by CorentinJ** [comment]
Sorry but I won't be available for 1 week. Take your time

**Comment by blue-fish** [comment]
No worries, I need the time anyway to train a model. I'll start working on the unified hyperparameters file and re-add you as a reviewer when I think it is ready for merge.

**Comment by CorentinJ** [comment]
Be careful with hparams, from my experience with big deep learning projects, no simple implementation of hparams will be a good all-round solution. I know the system I put in this repo is crappy, but beware that you are most likely opening a can of worms.

**Comment by blue-fish** [comment]
Can you elaborate a little bit more on that? Are you saying that for certain parameters, we might want to use different values for different parts of the model (encoder, synthesizer, vocoder)?

**Comment by CorentinJ** [comment]
If we keep the idea of them being constant or quasi constant through development, then it's fine.

If we start considering serializing hparams with model checkpoints, comparing hparams between models, adding backwards compatibility, ability to override etc... it can get very complex (whether merging in a single file or not).

**Comment by blue-fish** [comment]
Thank you for the clarification. For the "constant" case is there much value to collect them in the same file? It might be nice for ensuring compatibility between the synth and vocoder, but you already took care of that by [importing the relevant hparams](https://github.com/CorentinJ/Real-Time-Voice-Cloning/blob/a32962bb7b4827660646ac6dabf62309aea08a91/vocoder/hparams.py#L4-L16) from the synthesizer.

I lack the expertise to implement anything more sophisticated, so I might leave things as they are if there is not much to be gained by using a single hparams file.

**Comment by blue-fish** [comment]
Having spent several hours to get the Swedish model (#257) to work, I think it is a good idea to save the hparams along with the models. Maybe even load them at run time from the model files.

Then we can mix and match in the toolbox, and it can check for compatibility, e.g. `speaker_embedding_size` in encoder and synth, `sample_rate` between the synth and vocoder. Then we can make helpful error messages to replace the python exceptions that occur when models are incompatible.

Since we have to re-release models when this gets merged to master, it is an opportunity to implement this new checkpoint format. In addition to "model_state" and "optimizer_state", we can also save a dictionary of "model_parameters" which would contain something like:

#### Encoder
* `sample_rate`
* `speaker_embedding_size`
#### Synthesizer
* symbols
* language?
* `speaker_embedding_size`
* `sample_rate`, `n_mels`, `n_fft`, etc.
#### Vocoder
* `sample_rate`, `n_mels`, `n_fft`, etc.

I'll put this in a new issue and see if anyone wants this feature.

**Comment by blue-fish** [comment]
For anyone who is following this: I decided against including hparams along with the model checkpoints at this time. I just think the effort can be better spent elsewhere. Also don't see much benefit to consolidating the hparams in a single file, so will not be doing that either.

I am having trouble producing a pretrained model that can equal the current synthesizer. Although I generally don't have gaps, the cloned voice has some unpleasant distortion (like #314). Also, at my level of compute power I think removing all punctuation will help achieve a usable model faster. I'll be adding an option to ignore punctuation in training and/or inference.

The synthesized spectrograms from the fatchord tacotron are always identical. I might not be activating the dropout at inference time. Expect it should take a few days to resolve all of these.

**Comment by CorentinJ** [comment]
> Having spent several hours to get the Swedish model (#257) to work, I think it is a good idea to save the hparams along with the models. Maybe even load them at run time from the model files.
> 
> Then we can mix and match in the toolbox, and it can check for compatibility, e.g. `speaker_embedding_size` in encoder and synth, `sample_rate` between the synth and vocoder. Then we can make helpful error messages to replace the python exceptions that occur when models are incompatible.
> 
> Since we have to re-release models when this gets merged to master, it is an opportunity to implement this new checkpoint format. In addition to "model_state" and "optimizer_state", we can also save a dictionary of "model_parameters" which would contain something like:
> 
> #### Encoder
> * `sample_rate`
> * `speaker_embedding_size`
> 
> #### Synthesizer
> * symbols
> * language?
> * `speaker_embedding_size`
> * `sample_rate`, `n_mels`, `n_fft`, etc.
> 
> #### Vocoder
> * `sample_rate`, `n_mels`, `n_fft`, etc.
> 
> I'll put this in a new issue and see if anyone wants this feature.

This is exactly what you shouldn't do, and it's good that you gave up on it. 

For example, say you start training the voice encoder with default hparams. Then you train the synthesizer, and you decide to change some unrelated hparam in the synthesizer. You're already running into trouble because your hparams won't match that of the voice encoder checkpoint. In another scenario, you have a pretrained synthesizer model you want to finetune on another dataset, but the dataset requires different hparams than the one the synthesizer was trained on. Clash again. 

To work around that issue you would have to say which hparams are relevant to which model/situation, and which hparams should be compared and which should not. It's far from trivial. You also have to think about how the user inputs hparams in the CLI, and how the overrides happen. Does a model apply hparam overrides when you load it in memory? What if you have two instances of a model with different hparams?

As I said, better to have a simplistic but limited solution than to try to solve this. Not that there isn't a good solution, only that it takes a major development effort.

Regarding your synthesizer, is it expected to generate exactly the same output as the current tf one? If yes, make sure of it. If no, you'll have to retrain a vocoder.

**Comment by blue-fish** [comment]
With the addition of vocoder preprocess, the pytorch synthesizer fork is fully functional. It also has a pretrained model that can be used for testing purposes: https://github.com/CorentinJ/Real-Time-Voice-Cloning/issues/449#issuecomment-674420484

Rayhane has a lot of hparams! To simplify things I made assumptions for things that I don't expect most users to change, for example:
* symmetric_mels = True
* signal_normalization = True
* allow_clipping_in_normalization = True
* Generate GTA mels for vocoder preprocess

The code should be stable now, but I won't mark it as ready for review until I've retested all the functions. I expect that the lack of a suitable pretrained model is going to hold up the merge. @CorentinJ Is there any chance you can lend some compute power to training a new model on LibriSpeech/TTS if we provide the settings?

**Comment by CorentinJ** [comment]
> @CorentinJ Is there any chance you can lend some compute power

Sorry but I can't

**Comment by blue-fish** [comment]
@CorentinJ Do you mind if I eliminate the feature to override hparams at the command line? It takes a lot of code: https://github.com/tensorflow/tensorflow/blob/r1.15/tensorflow/contrib/training/python/training/hparam.py

**Comment by CorentinJ** [comment]
Where does it take a lot of code? Is it problematic to have that code?

**Comment by blue-fish** [comment]
> Where does it take a lot of code? Is it problematic to have that code?

Figured out a solution to this. I implemented a simple hparams override which uses `ast.literal_eval()` to convert the string representation of the values into other data types. Unlike the tensorflow class, I do not check that the overrides have the correct data type or even match to a valid key.

**Comment by blue-fish** [comment]
Ready for code review. There is a pretrained model for testing purposes: https://github.com/CorentinJ/Real-Time-Voice-Cloning/issues/449#issuecomment-674420484

**Comment by CorentinJ** [comment]
Got it, I'll do the review when I have time. I'll try to do it within a week

**Comment by blue-fish** [comment]
@CorentinJ I pushed and reverted a commit that overwrites the tacotron model with fatchord's version in WaveRNN. It lets you see the changes that were made to adapt it for SV2TTS: https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472/commits/ed1428142195eda87c0b3f6a8c60d69f1e78099e

File: synthesizer/inference.py
Comment (CorentinJ) [inline synthesizer/inference.py:109]: Why's it feeding inputs one by one instead of leveraging batching?

File: synthesizer/inference.py
Comment (CorentinJ) [inline synthesizer/inference.py:115]: You can remove all low mem functionalities if we can just run on CPU instead

File: synthesizer/models/tacotron.py
Comment (CorentinJ) [inline synthesizer/models/tacotron.py:325]: Should those be hardcoded or can you retrieve the values from elsewhere?

File: synthesizer/synthesize.py
Comment (CorentinJ) [inline synthesizer/synthesize.py:75]: Let's put slightly more informative names for x, m and e here

File: synthesizer/synthesize.py
Comment (CorentinJ) [inline synthesizer/synthesize.py:87]: pathlib

**Comment by blue-fish** [comment]
> Regarding pretrained models, what's your status? We can't merge unless we have a working synthesizer and a vocoder that are least on par with the previous model.

Still don't have a good model. Over the last 3 weeks I learned a little about what works and doesn't work.

VCTK trains fast and the attention mechanism works well with it. It's possible to synthesize very long texts without gaps. The downside to VCTK is the limited number of speakers which results in the cloned voice having lower similarity. In synthesized speech, phonemes are not pronounced consistently owing to the accent diversity within the dataset. VCTK also has a limited vocabulary which requires more input preprocessing (text cleaners) to achieve good pronunciation.

LibriSpeech/TTS results in gaps and is more likely to expose the weakness of fatchord's tacotron1 which is the lack of stop token prediction. It stops synthesizing whenever empty frames are generated, which occasionally leads to sentences being cut off prematurely when the speaker and/or words in the input text are unseen. After opening the PR I found a pytorch version of Rayhane's synthesizer (https://github.com/begeekmyfriend/tacotron2). We could try bringing it in and transferring the weights, and if that fails, train a tacotron2 model from scratch. 

> Since you're actively maintaining the repo, it could be less frustrating for you if you could contact me on an instant messaging app. Shout me [an email](corentin.jemine@gmail.com) if you want to do that.

The current arrangement works well enough for me, I'll let you know if that changes.

File: synthesizer/LICENSE.txt
Comment (blue-fish) [reply synthesizer/LICENSE.txt:3]: We are still using these portions of Rayhane's code:
* About 10 lines of code from [preprocess.py](https://github.com/Rayhane-mamah/Tacotron-2/blob/ab5cb08a931fc842d3892ebeb27c8b8734ddd4b8/preprocess.py#L20-L32)
* All of [utils/plot.py](https://github.com/Rayhane-mamah/Tacotron-2/blob/master/tacotron/utils/plot.py)
* The text utils which originate in keithito's tacotron repo
* Much of [datasets/audio.py](https://github.com/Rayhane-mamah/Tacotron-2/blob/master/datasets/audio.py) which is modified from keithito's tacotron repo

File: synthesizer/inference.py
Comment (blue-fish) [reply synthesizer/inference.py:109]: Short answer: That's how fatchord does it.

Longer answer: I tried to avoid making unnecessary changes to fatchord's code, figuring that cleanups and performance enhancements could come later. I implemented batching and didn't push it when I found that it didn't speed up inference. However, that was on CPU and I need to retest on GPU.

The synthesizer is noticeably slower in pytorch than tensorflow. Haven't formally benchmarked it but the existing TF model is about 2-3x faster for longer, multi-line inputs.

File: synthesizer/synthesize.py
Comment (blue-fish) [reply synthesizer/synthesize.py]: Fixed in https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472/commits/9187c25c20e7f31cbc86ae29aef1b0ec25e7cb94

File: synthesizer/synthesize.py
Comment (blue-fish) [reply synthesizer/synthesize.py]: Fixed in https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472/commits/9187c25c20e7f31cbc86ae29aef1b0ec25e7cb94

File: synthesizer/inference.py
Comment (blue-fish) [reply synthesizer/inference.py]: Removed low_mem in https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472/commits/70127563115ba71f73ab7a339024f8e46b107cfa , I also added a "--cpu" option to force the use of CPU when a low memory GPU is present.

File: synthesizer/models/tacotron.py
Comment (blue-fish) [reply synthesizer/models/tacotron.py]: Resolved in https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472/commits/53eb1d36fdf5f988b754a4e2d7a8443c7b3c0566 . It looks like the [256, 80] comes from Table 1 of [1703.10135](https://arxiv.org/pdf/1703.10135.pdf):

tacotron_table

`[256, 80]` can be related to existing hparams with `[postnet_dims*2, fft_bins]` but there might be a better way.

File: synthesizer/inference.py
Comment (blue-fish) [reply synthesizer/inference.py]: Added input batching for inference in https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472/commits/0ba3ba36dba44d000598b17687bd54ca8de52721

The resulting code is not as readable, but it might be faster.

**Comment by blue-fish** [comment]
The code issues identified should be resolved.

Regarding a pretrained model, I just started training on LibriSpeech `train-clean-100` and `train-clean-360`. With only 4gb of vram, my GPU is likely to have trouble finding the gradient as training progresses.

As mentioned earlier, we might need to add stop token prediction, else the synthesizer will terminate prematurely as it sometimes predicts empty frames whenever there is a failure to align. I'm going to look into implementing tacotron2 and transferring the existing weights, but I no longer have as much time to devote to this project as I did in recent weeks.

**Comment by blue-fish** [comment]
The LibriSpeech model prematurely stops synthesizing, but it is not because the decoder loop is stopped upon predicting an empty frame. Not sure if this is due to the attention mechanism (it uses Location-Sensitive Attention) because the alignment is robust right until it reaches a problem point in the input text and then the attention drops suddenly and completely. Will see if this gets better with additional training.

**Comment by blue-fish** [comment]
Trying the following:
1. Train LibriSpeech using the other attention mechanism included with fatchord's tacotron. (I think it is Bahdanau attention)
2. Start with a model trained on VCTK, then finetune on LibriSpeech. Do not allow update of weights associated with Location-Sensitive Attention and possibly the encoder as well. (Already tried it without the last part and it did not work well)

I was very careful about making changes to the model (see https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472#issuecomment-677730453) and implementing the training script. Since it is possible to train a model with robust attention on VCTK, I would not blame fatchord's tacotron1 entirely. The issue could be the quality of the training data (LibriSpeech/TTS) which may require us to decay the learning rate even faster.

**Comment by blue-fish** [comment]
> The LibriSpeech model prematurely stops synthesizing, but it is not because the decoder loop is stopped upon predicting an empty frame.

Solved this by trimming silences when preprocessing data, see #501 for the general idea. I am finding that fatchord's taco1 is more sensitive to silence than Rayhane's taco2, so I will make the modifications to the preprocessing scripts in this PR.

Still need to improve voice similarity. To get things to train faster, I cut the model's trainable parameters in half (text encoder dims = 256, decoder LSTM dims = 512, when they are double that in this repo). That may have been too much.

**Comment by blue-fish** [comment]
Voice quality improves by increasing the postnet dims from 128 to 512. It really bloats the model size, so I might run a test to see if 256 is enough. Similarity of the generated voice to the reference audio is on par the current synthesizer.

Refer to https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472#discussion_r479706745 . Also changing the postnet CBHG projection from a hardcoded `[256, 80]` to `[postnet_dims*2, fft_bins] = [512*2, 80]` adds a lot of trainable parameters and its value is uncertain. I'll study that code again, and possibly make that first element its own hparam.

I'm fairly satisfied with fatchord's taco1 now that I've figured out how to best work with it. It trains very quickly. Due to my limited memory I cut `max_mel_frames` from 900 to 300 (that is, 11.25 to 3.75 seconds) and found it still generalizes to longer inputs.

Still trying to work out solutions to these 2 problems.
1. Occasionally, inference stops prematurely. It correlates with words like `and` or `or`, or highly nonstandard text inputs (sequences of numbers mixed with text). This seems to be a problem of the training data and not the tacotron, but it's annoying. I'll run an experiment to see if switching from location-sensitive attention to forward attention improves it. Next thing to try is to increase the max_mel_frames.
2. Possibly due to bad prosody or silences in the input data, some embeddings result in synthesized utterances that contain numerous and unexpected pauses every few words.

In my opinion, these are less serious than the problems with the existing synthesizer (inability to synthesize short utterances, random long gaps). So if that's the bar we need to clear, then we could have a releasable pretrained model regardless of whether these problems get solved.

**Comment by CorentinJ** [comment]
Yes, lowering max mel frames is good, 500 or 700 is a good standard, 300 might be a little too small.

I'd still recommend another attention mechanism if you manage to get it working

**Comment by blue-fish** [comment]
### Samples and brief discussion of postnet

Voice quality is similar to the current synthesizer using 256 postnet_dims, but 512 has the potential for more depth. In this [**samples.zip**](https://github.com/CorentinJ/Real-Time-Voice-Cloning/files/5178474/samples.zip) you will find a bunch of pytorch samples at 256, and one at 512. The tensorflow samples are generated with the original pretrained models using default hparams. Tensorflow has somewhat better voice similarity, but the pytorch model only trained for 6 hours on a basic GPU so it's not a fair comparison.

Can confirm the prosody issue (frequent pauses for certain speaker embeddings) goes away when I train on a curated subset of LibriSpeech with no silences.

### An explanation for premature stop to inference?

The premature stop to inference is bugging me. Since no one else opened an issue for it in fatchord's repo, I started troubleshooting and think this is responsible:

https://github.com/blue-fish/Real-Time-Voice-Cloning/blob/14c3646e1737d0a71b7f12077fba38dcd1eee4ef/synthesizer/models/tacotron.py#L167

Our encoder output (size 256) is concat with the speaker embedding (size 256) as it goes into the decoder, which sees an input of size 512. The decoder prenet projects this down to 256, then 128 dimensions before passing it to the attention module. That might work for single-speaker (256 --> 128) but it's too much of an information bottleneck for voice cloning (512 --> 128). I'll increase the second linear projection to 256 to match this repo, and put it in hparams so it is not hardcoded.

We can also try leaving out the speaker embedding when calculating attention scores, since those should not depend on the speaker.

**Comment by blue-fish** [comment]
> Our encoder output (size 256) is concat with the speaker embedding (size 256) as it goes into the decoder, which sees an input of size 512. The decoder prenet projects this down to 256, then 128 dimensions before passing it to the attention module. That might work for single-speaker (256 --> 128) but it's too much of an information bottleneck for voice cloning (512 --> 128). I'll increase the second linear projection to 256 to match this repo, and put it in hparams so it is not hardcoded.

That actually made things worse. The information bottleneck is needed for the model to learn attention.

**Comment by blue-fish** [comment]
I found that the alignments are more robust when we don't project the encoder output when computing the attention scores. With commit https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472/commits/39c5943275f2a57009962e3689d58bcfdd9ec648 there is much less quitting during inference.

I still think the speaker embedding should be excluded from the inputs to the attention mechanism. This would be a difference from the SV2TTS paper, but if it completely solves the quitting problem then I am all for it.

**Comment by blue-fish** [comment]
With that last change the code revisions are complete, just need to train a model. My GPU is tied up performing other experiments for the next week, then I'll get back to it. I'll request another code review when testing is done and a good pretrained model is available.

**Comment by blue-fish** [comment]
Here is a pretrained model that can be used with the synthesizer: https://www.dropbox.com/s/gemy4hlm3fdf76u/librispeech_130k.zip?dl=0

1. Place `pretrained.pt` in `synthesizer/saved_models/pretrained/pretrained.pt`
2. Overwrite `synthesizer/hparams.py` with the one in the zip file.

Would anyone testing this please take the time to leave a comment with your thoughts on how well this pretrained model performs?

**Comment by blue-fish** [comment]
I need feedback from the community on whether this implementation is suitable, or if we need to try something different.

Personally, I find the attention failures annoying when they occur. Better attention mechanism and switching to tacotron2 may resolve this. I'm also experimenting with the so-called [Forward Tacotron](https://github.com/as-ideas/ForwardTacotron) which does not use any attention, but requires alignments for training.

Reproducible training of the pretrained model is also a goal. This tacotron1 is hard to train multispeaker without clean data, and until I gain more experience it's hard to tell if it's specific to this implementation or a more general issue with neural TTS.

**Comment by CorentinJ** [comment]
I just saw your update now, I'll be comparing your model with the base one sometime this week. It's been a while I haven't run the repo.

**Comment by CorentinJ** [comment]
Hey sorry for the delay. I got around testing a small sample with the demo but not with the toolbox. The toolbox crashes on start for me, that should be fixed. The sample I got was pretty bad unfortunately, but I'd like to test again with the toolbox working.

**Comment by blue-fish** [comment]
> The toolbox crashes on start for me

Sounds like #504 , try using matplotlib<=3.2.2 for now. Also, I pushed a new hparams.py, please use that instead of the one bundled with the pretrained model.

**Comment by blue-fish** [comment]
Also pushed a fix for the toolbox crashing on startup (presuming https://github.com/CorentinJ/Real-Time-Voice-Cloning/issues/504#issuecomment-678733034 is the cause of the crash).

**Comment by CorentinJ** [comment]
That fix works, thanks.

Both models are pretty bad so it's not so easy to judge, although I'd say the TF one is still doing better...
[Tensorflow](https://puu.sh/GA6fz.flac)
[PyTorch](https://puu.sh/GA6fm.flac)
[Reference audio](https://puu.sh/GA6fI.wav)

It's is annoying, I'd really like to get this merged in, but we can't afford to sacrifice more quality. What are your next steps for improving the model? Also, I can't remember whether the pretrained vocoder is trained on GT or GTA mels. GT would be best, to avoid depending on a given synthesizer.

**Comment by blue-fish** [comment]
I also concluded TF is still better than PT, based on a test of 6 speakers in https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472#issuecomment-687651582 . I'm a fan of fatchord's tacotron1 for its simplicity, but have yet to get good results for multispeaker despite a lot of experimentation. I plan to abandon it given the troubles I've had in getting it to train well.

The quality should improve by switching to tacotron2, however I do not have a timeline for when that might be completed.

**Comment by rallandr** [comment]
> Here is a pretrained model that can be used with the synthesizer: https://www.dropbox.com/s/gemy4hlm3fdf76u/librispeech_130k.zip?dl=0
> 
> 1. Place `pretrained.pt` in `synthesizer/saved_models/pretrained/pretrained.pt`
> 2. Ignore the hparams.py in the zip file. The required changes were merged into the branch in commit [`39f90da`](https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472/commits/39f90daa6ceb853657dace53a4449ffaf66cf349)
> 
> Would anyone testing this please take the time to leave a comment with your thoughts on how well this pretrained model performs?

Thanks for the material to start a Pytorch synthesizer, however I dont get how you make the toolbox run with this new format (.pt) for the syntesizer model. When using the toolbox it expects synthesizer weights... Thanks for any insight you could give

**Comment by blue-fish** [comment]
@rallandr You need to download the code from my fork to use this model: [`blue-fish:447_pytorch_synthesizer`](https://github.com/blue-fish/Real-Time-Voice-Cloning/tree/447_pytorch_synthesizer)

**Comment by rallandr** [comment]
Thanks a lot @blue-fish

**Comment by rallandr** [comment]
Just so you know, there is an issue which I think is specific to Windows when trying to train the synthesizer, I think it's related to multiprocessing and for some reason it doesn't work on Windows...

Traceback (most recent call last):
  **File "synthesizer_train.py", line 35, in <module>
    train(**vars(args))
  File "D:\dfvoc\Real-Time-Voice-Cloning-447_pytorch_synthesizer\synthesizer\train.py", line 158, in train
    for i, (texts, mels, embeds, idx) in enumerate(data_loader, 1):**
  File "C:\Users\Remi\AppData\Local\Programs\Python\Python37\lib\site-packages\torch\utils\data\dataloader.py", line 359, in __iter__
    return self._get_iterator()
  File "C:\Users\Remi\AppData\Local\Programs\Python\Python37\lib\site-packages\torch\utils\data\dataloader.py", line 301, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
  File "C:\Users\Remi\AppData\Local\Programs\Python\Python37\lib\site-packages\torch\utils\data\dataloader.py", line 885, in __init__
    w.start()
  File "C:\Users\Remi\AppData\Local\Programs\Python\Python37\lib\multiprocessing\process.py", line 112, in start
    self._popen = self._Popen(self)
  File "C:\Users\Remi\AppData\Local\Programs\Python\Python37\lib\multiprocessing\context.py", line 223, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
  File "C:\Users\Remi\AppData\Local\Programs\Python\Python37\lib\multiprocessing\context.py", line 322, in _Popen
    return Popen(process_obj)
  File "C:\Users\Remi\AppData\Local\Programs\Python\Python37\lib\multiprocessing\popen_spawn_win32.py", line 89, in __init__
    reduction.dump(process_obj, to_child)
  File "C:\Users\Remi\AppData\Local\Programs\Python\Python37\lib\multiprocessing\reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
**AttributeError: Can't pickle local object 'train.\<locals>.\<lambda>'**
PS D:\dfvoc\Real-Time-Voice-Cloning-447_pytorch_synthesizer> Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\Remi\AppData\Local\Programs\Python\Python37\lib\multiprocessing\spawn.py", line 105, in spawn_main
    exitcode = _main(fd)
  File "C:\Users\Remi\AppData\Local\Programs\Python\Python37\lib\multiprocessing\spawn.py", line 115, in _main
    self = reduction.pickle.load(from_parent)
**EOFError: Ran out of input**

I'll try yo find a way around it

File: demo_cli.py
Comment (Kayceelynn84) [inline demo_cli.py:30]: T

File: demo_cli.py
Comment (Kayceelynn84) [inline demo_cli.py:45]: - ****_[@]()_****[]()

**Review by Kayceelynn84 — APPROVED** [review]

**Comment by blue-fish** [comment]
(Deleted spam comments from pull request)

**Comment by blue-fish** [comment]
My tensorflow synth from scratch in #538 has similar voice quality issues as the [pretrained model](https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472#issuecomment-695206377) shared here. So I think the lower quality is not due to the tacotron 1/2 difference, but instead by:
1. **Smaller batch size:** I used batch 12 instead of 144 due to limited vram, and/or
2. **Smaller training dataset:** I curated LibriSpeech aggressively to remove gaps

**Review by fxmorin — APPROVED** [review]

**Comment by blue-fish** [comment]
The attention issues can be resolved by matching the settings of `LocationSensitiveAttention` in Rayhane's tacotron. It's also necessary to use a stop token prediction since attention is less stable for multispeaker. Silent mel frames are common, which causes fatchord's Taco1 to prematurely stop generation. Adding a stop token forces it to finish. With attention fixed, training on the full LibriSpeech dataset yields a usable model with voice quality comparable to tensorflow.

I'll push the code changes when the pretrained model is ready, to avoid disrupting those actively using this branch.

**Comment by blue-fish** [comment]
Attention is problematic when the speaker embedding is generated from a synthetic utterance. The current synth model is more robust in this regard. I will experiment and see if this can be improved by replacing the attention GRU with a 2-layer LSTM as in Tacotron2.

**Comment by blue-fish** [comment]
This line seems to be the culprit of the attention issues: [synthesizer/models/tacotron.py#L227](https://github.com/blue-fish/Real-Time-Voice-Cloning/blob/44bb76561499c16222b9636167322e286de978b1/synthesizer/models/tacotron.py#L227)

I've updated it to match Rayhane's LSA, which only uses the cumulative attention score in the location convolution. I don't know why fatchord does the concat with the previous attention output. It learns faster, but is less stable for unseen speakers.

**Comment by blue-fish** [comment]
The pytorch model has attention failures when the speaker embedding is generated from a synthetic utterance or low-quality recording. I still don't understand why this is the case. Audio quality is just as good as tensorflow when it does work, but it's not releasable since it's inferior model to what we already have.

I've spent a lot of time debugging and experimenting with fatchord's model, without success. It is time for me to move on. For those that would like to continue working on this, I've pushed all my work to the branch. However, switching to a proven tacotron2 implementation would be best.

The training and inference code developed for this PR can be reused for a pytorch-based synthesizer.

**Comment by blue-fish** [comment]
I notice that Rayhane's attention implementation masks text padding, while fatchord's does not. This seems like a plausible explanation for the attention difficulties. I'll try adding a mask and see if it improves the results.

**Comment by blue-fish** [comment]
Training is still in progress. Here are some observations:
* Tacotron 1 learns faster than Tacotron 2. Training time per step is about the same, but the model is usable in fewer steps. Voice quality is similar. Attention is more robust in Taco2.
* Voice quality improves when training on the full LibriSpeech dataset instead of the no-gaps subset described in #538. It sounds quite similar to Corentin's model even with a batch size of 12 instead of 144.
* Silence in the middle of training data causes long gaps, skipped words, and stuttering. I tried solving this by trimming silences with webrtcvad. This trims long gaps into very short gaps, which results in random pauses between words.

**Comment by blue-fish** [comment]
Development of the pytorch synthesizer is complete. Please review the changes.

The [pretrained model release](https://github.com/blue-fish/Real-Time-Voice-Cloning/releases/tag/v1.0) consists of the pretrained encoder, along with synthesizer and vocoder models I have developed. Audio samples and model details: https://blue-fish.github.io/experiments/RTVC-7.html

**Comment by CorentinJ** [comment]
Wow, amazing work. I'll do my best to find the time to review within this week.

**Comment by Garvit-32** [comment]
Hi @blue-fish why you are using tacotron1 instead of tacotron2?  Can we change this code from tacotron1 to tacotron2 in pytorch?

**Comment by blue-fish** [comment]
@Garvit-32
The main reason to use Taco1 is that it uses the same codebase with the vocoder (fatchord/WaveRNN). The commonality makes it a lot easier to write the training script and integrate it with the rest of the repo. Now that the supporting infrastructure is in place, the model can be switched with relative ease.

I've already shared my thoughts on Taco1 vs Taco2 in https://github.com/CorentinJ/Real-Time-Voice-Cloning/pull/472#issuecomment-769914574 . They're close in performance. I prefer Taco2.

**Review by CorentinJ — APPROVED** [review]
Alright well I checked the code and played around with the toolbox. I think all seems good.

Thank you again for your amazing work, feel free to merge whenever.

**Comment by 121898** [comment]
> Development of the pytorch synthesizer is complete. Please review the changes.
> 
> The [pretrained model release](https://github.com/blue-fish/Real-Time-Voice-Cloning/releases/tag/v1.0) consists of the pretrained encoder, along with synthesizer and vocoder models I have developed. Audio samples and model details: https://blue-fish.github.io/experiments/RTVC-7.html

Hi, thanks for the amazing work, may I know what does Google rows mean in your demo page? Thanks

**Comment by blue-fish** [comment]
@121898 Google rows are the audio samples from 1806.04558. https://google.github.io/tacotron/publications/speaker_adaptation/index.html

**Pull request merged**
